cdap-encryption-ext-tink/src/main/java/io/cdap/cdap/encryption/tink/CloudKmsClient.java

```diff
@@ -0,0 +1,180 @@
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.time.temporal.ChronoUnit;
+import java.util.HashSet;
+import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Client wrapping Cloud KMS.
+ */
+public class CloudKmsClient implements Closeable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CloudKmsClient.class);
+
+  private final KeyManagementServiceClient client;
+  private final String projectId;
+  private final String location;
+  private final String keyRingId;
+  // In-memory cache to hold created crypto keys, this is to avoid checking if a given crypto key exists.
+  private final Set<String> knownCryptoKeys;
+
+  /**
+   * Constructs Cloud KMS client.
+   *
+   * @throws IOException if cloud kms client can not be created
+   */
+  CloudKmsClient(String projectId, String location, String keyRingId) throws IOException {
+    this.projectId = projectId;
+    this.location = location;
+    this.keyRingId = keyRingId;
+    this.client = KeyManagementServiceClient.create();
+    this.knownCryptoKeys = new HashSet<>();
+  }
+
+  /**
+   * Get project id from the metadata server. Makes a request to the metadata server that lives on
+   * the VM, as described at https://cloud.google.com/compute/docs/storing-retrieving-metadata.
+   */
+  private String getSystemProjectId(String metadataServerApi) throws IOException {
+    URL url = new URL(metadataServerApi);
+    HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+    try {
+      connection = (HttpURLConnection) url.openConnection();
+      connection.setRequestProperty("Metadata-Flavor", "Google");
+      connection.connect();
+      try (Reader reader = new InputStreamReader(connection.getInputStream(),
+          StandardCharsets.UTF_8)) {
+        return CharStreams.toString(reader);
+      }
+    } finally {
+      connection.disconnect();
+    }
+  }
+
+  /**
+   * Creates a new key ring with the given id.
+   *
+   * @throws IOException if there's an error while creating the key ring
+   */
+  void createKeyRingIfNotExists() throws CipherInitializationException {
+    LOG.debug("Creating key ring with id {} in projects/{}/locations/{}.", keyRingId, projectId,
+        location);
+
+    try {
+      client.getKeyRing(KeyRingName.of(projectId, location, keyRingId));
+      // If we get here, the key ring was found, so we skip creating the key ring.
+      return;
+    } catch (ApiException e) {
+      if (!e.getStatusCode().getCode().equals(StatusCode.Code.NOT_FOUND)) {
+        throw new CipherInitializationException("Failed to check if Cloud KMS key ring exists", e);
+      }
+    }
+
+    LocationName locationName = LocationName.of(projectId, location);
+    client.createKeyRing(locationName, keyRingId, KeyRing.newBuilder().build());
```

Comment:
should catch ApiException here too. At least to check if it's a 409 which can be ignored, probably want to translate it to a CipherInitializationException for other errors

Reply:
Done